File: ext/internal/socket/message_state.go
package socket

import (
	"bufio"
	"errors"
	"io"
	"net"
	"sconcur/internal/contracts"
	"sconcur/internal/dto"
	"sconcur/internal/errs"
	"sconcur/internal/helpers"
	"sync"
	"time"
)

var _ contracts.StateContract = (*MessageState)(nil)

// MessageState streams the inbound length-prefixed frames of one connection to PHP,
// one frame per Next(). Shared by the server (the handler reads via
// Connection::read()) and the client (the same, dial-side). Implements
// contracts.StateContract.
type MessageState struct {
	mutex           sync.Mutex
	message         *dto.Message
	conn            net.Conn
	reader          *bufio.Reader
	readTimeout     time.Duration
	maxMessageBytes int
	errFactory      *errs.Factory
	startTime       time.Time
}

// NewMessageState builds the inbound frame stream for a connection. readTimeout 0
// means no idle deadline; maxMessageBytes 0 means no inbound frame size limit.
func NewMessageState(
	message *dto.Message,
	conn net.Conn,
	reader *bufio.Reader,
	readTimeout time.Duration,
	maxMessageBytes int,
	errFactory *errs.Factory,
) *MessageState {
	return &MessageState{
		message:         message,
		conn:            conn,
		reader:          reader,
		readTimeout:     readTimeout,
		maxMessageBytes: maxMessageBytes,
		errFactory:      errFactory,
		startTime:       time.Now(),
	}
}

func (s *MessageState) Next() *dto.Result {
	s.mutex.Lock()
	defer s.mutex.Unlock()

	if s.readTimeout > 0 {
		_ = s.conn.SetReadDeadline(time.Now().Add(s.readTimeout))
	} else {
		_ = s.conn.SetReadDeadline(time.Time{})
	}

	frame, err := ReadFrame(s.reader, s.maxMessageBytes)

	if err != nil {
		// A clean connection end (EOF, closed read side on drain, idle timeout):
		// end the stream so the PHP loop exits and the connection coroutine finishes
		// without surfacing an error to the handler.
		if isConnectionClosed(err) {
			return dto.NewSuccessResult(s.message, "", helpers.CalcExecutionMs(s.startTime))
		}

		return dto.NewErrorResult(s.message, s.errFactory.ByErr("read frame", err))
	}

	return dto.NewSuccessResultWithNext(s.message, string(frame), helpers.CalcExecutionMs(s.startTime))
}

func (s *MessageState) Close() {
	// The connection itself is closed by its owner (the server's handleConn or the
	// client's connection cleanup); nothing to do here.
}

// isConnectionClosed reports whether a read error means the connection has ended
// normally (peer closed, our own read side closed on drain, or the idle read
// deadline elapsed) — all of which finish the stream cleanly rather than as an
// error the handler must see.
func isConnectionClosed(err error) bool {
	if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) || errors.Is(err, net.ErrClosed) {
		return true
	}

	var netErr net.Error

	if errors.As(err, &netErr) && netErr.Timeout() {
		return true
	}

	return false
}

File: ext-rust/src/features/wsserver/message_state.rs
//! Mirrors ext/internal/features/wsserver/message_state.go: the inbound data
//! messages of one connection, streamed to PHP one message per `next()`.
//!
//! The connection is pumped by a dedicated read task, so control frames are
//! processed even when the handler is push-only and never reads; this state only
//! sees the data messages that task forwards.

use std::sync::Arc;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::{Duration, Instant};
use tokio::sync::mpsc;
use tokio_util::sync::CancellationToken;

use crate::dto::{Message, Result};
use crate::helpers::calc_execution_ms;
use crate::states::{StateCloseFuture, StateContract, StateFuture};
use crate::ws::{InboundMessage, encode_inbound};

/// How long `next()` still waits for the connection's FIRST inbound message
/// after the drain signal fires.
///
/// The limiting connection under a concurrency cap is dispatched and drained
/// almost simultaneously, so its first message may still be on the wire when the
/// drain begins; without this window the handler gets EOF and that exchange is
/// silently bounced. A connection that has already delivered a message keeps the
/// immediate EOF.
const FIRST_MESSAGE_DRAIN_GRACE: Duration = Duration::from_millis(250);

pub struct MessageState {
    message: Arc<Message>,
    messages: tokio::sync::Mutex<mpsc::Receiver<InboundMessage>>,
    drain: CancellationToken,
    delivered: AtomicU64,
    start_time: Instant,
}

impl MessageState {
    pub fn new(
        message: Arc<Message>,
        messages: mpsc::Receiver<InboundMessage>,
        drain: CancellationToken,
    ) -> Self {
        MessageState {
            message,
            messages: tokio::sync::Mutex::new(messages),
            drain,
            delivered: AtomicU64::new(0),
            start_time: Instant::now(),
        }
    }

    fn finished(&self) -> Result {
        Result::success(&self.message, Vec::new(), calc_execution_ms(self.start_time))
    }

    fn delivered(&self, message: InboundMessage) -> Result {
        self.delivered.fetch_add(1, Ordering::Relaxed);

        Result::success_with_next(
            &self.message,
            encode_inbound(&message),
            calc_execution_ms(self.start_time),
        )
    }
}

impl StateContract for MessageState {
    fn next(&self) -> StateFuture<'_> {
        Box::pin(async move {
            let mut messages = self.messages.lock().await;

            // A message that already arrived wins over the drain signal. With
            // both ready, a plain select would pick at random and could drop a
            // buffered message, answering EOF while data was waiting — the
            // half-close semantic is "stop new input, flush what arrived".
            match messages.try_recv() {
                Ok(message) => return self.delivered(message),
                Err(mpsc::error::TryRecvError::Disconnected) => return self.finished(),
                Err(mpsc::error::TryRecvError::Empty) => {}
            }

            if self.drain.is_cancelled() {
                if self.delivered.load(Ordering::Relaxed) > 0 {
                    return self.finished();
                }

                return match tokio::time::timeout(FIRST_MESSAGE_DRAIN_GRACE, messages.recv()).await
                {
                    Ok(Some(message)) => self.delivered(message),
                    _ => self.finished(),
                };
            }

            tokio::select! {
                biased;

                received = messages.recv() => match received {
                    Some(message) => self.delivered(message),
                    None => self.finished(),
                },
                _ = self.drain.cancelled() => {
                    // Same rule on the racing path: flush a message that landed
                    // with the signal rather than answering EOF over it.
                    match messages.try_recv() {
                        Ok(message) => self.delivered(message),
                        _ => self.finished(),
                    }
                }
            }
        })
    }

    fn close(&self) -> StateCloseFuture<'_> {
        Box::pin(async move {
            // The connection is closed by its owner (the server's connection
            // task); nothing to do here.
        })
    }
}
